File: src/pages/azure-ml-studio.ts
import { Page } from '@playwright/test';
import { BasePage } from './base-page';
import { TestLogger } from '../helpers/logger';

export class AzureMLStudioPage extends BasePage {
  // Selectors
  private readonly selectors = {
    // Navigation
    homeButton: '[data-testid="home-button"]',
    workspaceSelector: '[data-testid="workspace-selector"]',
    navigationMenu: '[data-testid="navigation-menu"]',
    
    // Workspace
    workspaceName: '[data-testid="workspace-name"]',
    workspaceOverview: '[data-testid="workspace-overview"]',
    
    // Compute
    computeTab: '[data-testid="compute-tab"]',
    computeInstances: '[data-testid="compute-instances"]',
    computeClusters: '[data-testid="compute-clusters"]',
    createComputeButton: '[data-testid="create-compute-button"]',
    
    // Notebooks
    notebooksTab: '[data-testid="notebooks-tab"]',
    notebooksList: '[data-testid="notebooks-list"]',
    createNotebookButton: '[data-testid="create-notebook-button"]',
    uploadNotebookButton: '[data-testid="upload-notebook-button"]',
    
    // Jobs
    jobsTab: '[data-testid="jobs-tab"]',
    jobsList: '[data-testid="jobs-list"]',
    createJobButton: '[data-testid="create-job-button"]',
    
    // Data
    dataTab: '[data-testid="data-tab"]',
    dataAssets: '[data-testid="data-assets"]',
    datastores: '[data-testid="datastores"]',
    
    // Models
    modelsTab: '[data-testid="models-tab"]',
    modelsList: '[data-testid="models-list"]',
    
    // Endpoints
    endpointsTab: '[data-testid="endpoints-tab"]',
    endpointsList: '[data-testid="endpoints-list"]',
    
    // Common elements
    loadingSpinner: '[data-testid="loading-spinner"]',
    errorMessage: '[data-testid="error-message"]',
    successMessage: '[data-testid="success-message"]',
    confirmDialog: '[data-testid="confirm-dialog"]',
    confirmButton: '[data-testid="confirm-button"]',
    cancelButton: '[data-testid="cancel-button"]',
  };

  constructor(page: Page, testLogger?: TestLogger) {
    super(page, testLogger);
  }

  async isPageLoaded(): Promise<boolean> {
    try {
      await this.waitForElement(this.selectors.workspaceName, 10000);
      return true;
    } catch {
      return false;
    }
  }

  getPageIdentifier(): string {
    return 'Azure ML Studio';
  }

  async waitForPageLoad(): Promise<void> {
    await super.waitForPageLoad();
    
    // Wait for workspace to load
    await this.waitForElement(this.selectors.workspaceName);
    
    // Wait for loading spinner to disappear
    try {
      await this.page.waitForSelector(this.selectors.loadingSpinner, { 
        state: 'detached', 
        timeout: 30000 
      });
    } catch {
      // Loading spinner might not appear for fast loads
    }
  }

  // Navigation methods
  async navigateToCompute(): Promise<void> {
    this.testLogger?.step('Navigating to Compute section');
    await this.clickElement(this.selectors.computeTab);
    await this.waitForElement(this.selectors.computeInstances);
  }

  async navigateToNotebooks(): Promise<void> {
    this.testLogger?.step('Navigating to Notebooks section');
    await this.clickElement(this.selectors.notebooksTab);
    await this.waitForElement(this.selectors.notebooksList);
  }

  async navigateToJobs(): Promise<void> {
    this.testLogger?.step('Navigating to Jobs section');
    await this.clickElement(this.selectors.jobsTab);
    await this.waitForElement(this.selectors.jobsList);
  }

  async navigateToData(): Promise<void> {
    this.testLogger?.step('Navigating to Data section');
    await this.clickElement(this.selectors.dataTab);
    await this.waitForElement(this.selectors.dataAssets);
  }

  async navigateToModels(): Promise<void> {
    this.testLogger?.step('Navigating to Models section');
    await this.clickElement(this.selectors.modelsTab);
    await this.waitForElement(this.selectors.modelsList);
  }

  async navigateToEndpoints(): Promise<void> {
    this.testLogger?.step('Navigating to Endpoints section');
    await this.clickElement(this.selectors.endpointsTab);
    await this.waitForElement(this.selectors.endpointsList);
  }

  // Workspace methods
  async getWorkspaceName(): Promise<string> {
    return await this.getElementText(this.selectors.workspaceName);
  }

  async selectWorkspace(workspaceName: string): Promise<void> {
    this.testLogger?.action(`Selecting workspace: ${workspaceName}`);
    
    await this.clickElement(this.selectors.workspaceSelector);
    await this.waitForText(workspaceName);
    await this.clickElement(`text=${workspaceName}`);
    
    // Wait for workspace to load
    await this.waitForPageLoad();
  }

  // Compute methods
  async getComputeInstances(): Promise<string[]> {
    await this.navigateToCompute();
    
    const instances = await this.page.locator(`${this.selectors.computeInstances} [data-testid="compute-instance-name"]`).all();
    const names = [];
    
    for (const instance of instances) {
      const name = await instance.textContent();
      if (name) names.push(name.trim());
    }
    
    return names;
  }

  async createComputeInstance(name: string, vmSize: string = 'Standard_DS3_v2'): Promise<void> {
    this.testLogger?.action(`Creating compute instance: ${name}`);
    
    await this.navigateToCompute();
    await this.clickElement(this.selectors.createComputeButton);
    
    // Fill compute instance form
    await this.fillInput('[data-testid="compute-name-input"]', name);
    await this.selectOption('[data-testid="vm-size-select"]', vmSize);
    
    // Submit form
    await this.clickElement('[data-testid="create-button"]');
    
    // Wait for creation to complete
    await this.waitForText(`${name}`);
    this.testLogger?.info(`Compute instance created: ${name}`);
  }

  async startComputeInstance(name: string): Promise<void> {
    this.testLogger?.action(`Starting compute instance: ${name}`);
    
    await this.navigateToCompute();
    
    // Find the compute instance row and click start button
    const instanceRow = this.page.locator(`[data-testid="compute-instance-row"]:has-text("${name}")`);
    await instanceRow.locator('[data-testid="start-button"]').click();
    
    // Wait for status to change to "Running"
    await this.waitForComputeInstanceStatus(name, 'Running');
    this.testLogger?.info(`Compute instance started: ${name}`);
  }

  async stopComputeInstance(name: string): Promise<void> {
    this.testLogger?.action(`Stopping compute instance: ${name}`);
    
    await this.navigateToCompute();
    
    // Find the compute instance row and click stop button
    const instanceRow = this.page.locator(`[data-testid="compute-instance-row"]:has-text("${name}")`);
    await instanceRow.locator('[data-testid="stop-button"]').click();
    
    // Confirm stop action
    await this.clickElement(this.selectors.confirmButton);
    
    // Wait for status to change to "Stopped"
    await this.waitForComputeInstanceStatus(name, 'Stopped');
    this.testLogger?.info(`Compute instance stopped: ${name}`);
  }

  async getComputeInstanceStatus(name: string): Promise<string> {
    await this.navigateToCompute();
    
    const instanceRow = this.page.locator(`[data-testid="compute-instance-row"]:has-text("${name}")`);
    const statusElement = instanceRow.locator('[data-testid="compute-status"]');
    
    return await statusElement.textContent() || 'Unknown';
  }

  async waitForComputeInstanceStatus(name: string, expectedStatus: string, timeout: number = 600000): Promise<void> {
    this.testLogger?.info(`Waiting for compute instance ${name} to reach status: ${expectedStatus}`);
    
    const startTime = Date.now();
    
    while (Date.now() - startTime < timeout) {
      const currentStatus = await this.getComputeInstanceStatus(name);
      
      if (currentStatus.toLowerCase() === expectedStatus.toLowerCase()) {
        this.testLogger?.info(`Compute instance ${name} reached status: ${expectedStatus}`);
        return;
      }
      
      this.testLogger?.debug(`Compute instance ${name} current status: ${currentStatus}, waiting...`);
      await this.page.waitForTimeout(10000); // Wait 10 seconds between checks
    }
    
    throw new Error(`Timeout waiting for compute instance ${name} to reach status ${expectedStatus}`);
  }

  // Notebook methods
  async getNotebooks(): Promise<string[]> {
    await this.navigateToNotebooks();
    
    const notebooks = await this.page.locator(`${this.selectors.notebooksList} [data-testid="notebook-name"]`).all();
    const names = [];
    
    for (const notebook of notebooks) {
      const name = await notebook.textContent();
      if (name) names.push(name.trim());
    }
    
    return names;
  }

  async createNotebook(name: string, kernelType: string = 'Python 3.8'): Promise<void> {
    this.testLogger?.action(`Creating notebook: ${name}`);
    
    await this.navigateToNotebooks();
    await this.clickElement(this.selectors.createNotebookButton);
    
    // Fill notebook form
    await this.fillInput('[data-testid="notebook-name-input"]', name);
    await this.selectOption('[data-testid="kernel-select"]', kernelType);
    
    // Submit form
    await this.clickElement('[data-testid="create-button"]');
    
    // Wait for notebook to be created and opened
    await this.waitForUrl(/.*notebooks.*/, 30000);
    this.testLogger?.info(`Notebook created: ${name}`);
  }

  async openNotebook(name: string): Promise<void> {
    this.testLogger?.action(`Opening notebook: ${name}`);
    
    await this.navigateToNotebooks();
    
    // Find and click the notebook
    const notebookRow = this.page.locator(`[data-testid="notebook-row"]:has-text("${name}")`);
    await notebookRow.click();
    
    // Wait for notebook to open
    await this.waitForUrl(/.*notebooks.*/, 30000);
    this.testLogger?.info(`Notebook opened: ${name}`);
  }

  async uploadNotebook(filePath: string): Promise<void> {
    this.testLogger?.action(`Uploading notebook: ${filePath}`);
    
    await this.navigateToNotebooks();
    await this.clickElement(this.selectors.uploadNotebookButton);
    
    // Upload file
    await this.uploadFile('[data-testid="file-upload-input"]', filePath);
    
    // Submit upload
    await this.clickElement('[data-testid="upload-button"]');
    
    // Wait for upload to complete
    await this.waitForElement(this.selectors.successMessage);
    this.testLogger?.info(`Notebook uploaded: ${filePath}`);
  }

  // Job methods
  async getJobs(): Promise<string[]> {
    await this.navigateToJobs();
    
    const jobs = await this.page.locator(`${this.selectors.jobsList} [data-testid="job-name"]`).all();
    const names = [];
    
    for (const job of jobs) {
      const name = await job.textContent();
      if (name) names.push(name.trim());
    }
    
    return names;
  }

  async createJob(jobConfig: any): Promise<void> {
    this.testLogger?.action('Creating new job');
    
    await this.navigateToJobs();
    await this.clickElement(this.selectors.createJobButton);
    
    // This would depend on the specific job creation UI
    // Implementation would vary based on job type and UI structure
    
    this.testLogger?.info('Job creation initiated');
  }

  async getJobStatus(jobName: string): Promise<string> {
    await this.navigateToJobs();
    
    const jobRow = this.page.locator(`[data-testid="job-row"]:has-text("${jobName}")`);
    const statusElement = jobRow.locator('[data-testid="job-status"]');
    
    return await statusElement.textContent() || 'Unknown';
  }

  // Assertion methods
  async assertWorkspaceLoaded(workspaceName: string): Promise<void> {
    const actualName = await this.getWorkspaceName();
    if (actualName !== workspaceName) {
      throw new Error(`Expected workspace "${workspaceName}" but got "${actualName}"`);
    }
    
    this.testLogger?.assertion(`Workspace loaded: ${workspaceName}`, true);
  }

  async assertComputeInstanceExists(name: string): Promise<void> {
    const instances = await this.getComputeInstances();
    if (!instances.includes(name)) {
      throw new Error(`Compute instance "${name}" not found. Available instances: ${instances.join(', ')}`);
    }
    
    this.testLogger?.assertion(`Compute instance exists: ${name}`, true);
  }

  async assertComputeInstanceStatus(name: string, expectedStatus: string): Promise<void> {
    const actualStatus = await this.getComputeInstanceStatus(name);
    if (actualStatus.toLowerCase() !== expectedStatus.toLowerCase()) {
      throw new Error(`Expected compute instance "${name}" status "${expectedStatus}" but got "${actualStatus}"`);
    }
    
    this.testLogger?.assertion(`Compute instance ${name} status is ${expectedStatus}`, true);
  }

  async assertNotebookExists(name: string): Promise<void> {
    const notebooks = await this.getNotebooks();
    if (!notebooks.includes(name)) {
      throw new Error(`Notebook "${name}" not found. Available notebooks: ${notebooks.join(', ')}`);
    }
    
    this.testLogger?.assertion(`Notebook exists: ${name}`, true);
  }

  async assertJobExists(name: string): Promise<void> {
    const jobs = await this.getJobs();
    if (!jobs.includes(name)) {
      throw new Error(`Job "${name}" not found. Available jobs: ${jobs.join(', ')}`);
    }
    
    this.testLogger?.assertion(`Job exists: ${name}`, true);
  }
}

File: src/azure_ml_automation/pages/azure_ml_studio.py
"""Azure ML Studio page object."""

from typing import List, Optional

from playwright.async_api import Page

from .base_page import BasePage
from ..helpers.config import config
from ..helpers.logger import TestLogger


class AzureMLStudioPage(BasePage):
    """Page object for Azure ML Studio."""
    
    # Selectors
    WORKSPACE_SELECTOR = "[data-testid='workspace-selector']"
    COMPUTE_NAV_ITEM = "[data-testid='nav-compute']"
    NOTEBOOKS_NAV_ITEM = "[data-testid='nav-notebooks']"
    JOBS_NAV_ITEM = "[data-testid='nav-jobs']"
    DATASETS_NAV_ITEM = "[data-testid='nav-datasets']"
    
    # Compute page selectors
    COMPUTE_INSTANCES_TAB = "[data-testid='compute-instances-tab']"
    COMPUTE_CLUSTERS_TAB = "[data-testid='compute-clusters-tab']"
    CREATE_COMPUTE_BUTTON = "[data-testid='create-compute-button']"
    COMPUTE_INSTANCE_NAME_INPUT = "[data-testid='compute-name-input']"
    COMPUTE_SIZE_DROPDOWN = "[data-testid='compute-size-dropdown']"
    CREATE_BUTTON = "[data-testid='create-button']"
    
    # Compute instance actions
    START_COMPUTE_BUTTON = "[data-testid='start-compute-button']"
    STOP_COMPUTE_BUTTON = "[data-testid='stop-compute-button']"
    DELETE_COMPUTE_BUTTON = "[data-testid='delete-compute-button']"
    APPLICATIONS_BUTTON = "[data-testid='applications-button']"
    VSCODE_DESKTOP_BUTTON = "[data-testid='vscode-desktop-button']"
    
    # Status indicators
    COMPUTE_STATUS_RUNNING = "[data-testid='status-running']"
    COMPUTE_STATUS_STOPPED = "[data-testid='status-stopped']"
    COMPUTE_STATUS_CREATING = "[data-testid='status-creating']"
    
    # Notebook selectors
    NOTEBOOK_LIST = "[data-testid='notebook-list']"
    NEW_NOTEBOOK_BUTTON = "[data-testid='new-notebook-button']"
    UPLOAD_NOTEBOOK_BUTTON = "[data-testid='upload-notebook-button']"
    NOTEBOOK_CELL = "[data-testid='notebook-cell']"
    RUN_CELL_BUTTON = "[data-testid='run-cell-button']"
    RUN_ALL_BUTTON = "[data-testid='run-all-button']"
    
    # Job selectors
    JOB_LIST = "[data-testid='job-list']"
    CREATE_JOB_BUTTON = "[data-testid='create-job-button']"
    JOB_STATUS = "[data-testid='job-status']"
    
    def __init__(self, page: Page, test_logger: Optional[TestLogger] = None):
        super().__init__(page, test_logger)
        self.base_url = config.urls.base
    
    async def is_page_loaded(self) -> bool:
        """Check if Azure ML Studio is loaded."""
        try:
            # Check for multiple possible indicators that the page has loaded
            selectors_to_check = [
                self.WORKSPACE_SELECTOR,
                "[data-testid*='workspace']",
                "[class*='workspace']",
                "h1, h2, h3",  # Any heading
                "[role='main']",  # Main content area
                "nav",  # Navigation
                ".ms-Nav",  # Microsoft navigation component
                "[data-automation-id]"  # Any automation ID
            ]
            
            for selector in selectors_to_check:
                try:
                    await self.wait_for_element(selector, timeout=2000)
                    if self.test_logger:
                        self.test_logger.debug(f"Page loaded - found element: {selector}")
                    return True
                except Exception:
                    continue
            
            # If no specific selectors found, check if page title contains Azure ML
            try:
                title = await self.page.title()
                if any(keyword in title.lower() for keyword in ['azure', 'ml', 'machine learning', 'studio']):
                    if self.test_logger:
                        self.test_logger.debug(f"Page loaded - title check passed: {title}")
                    return True
            except Exception:
                pass
            
            # Final fallback - check if page is not loading
            try:
                await self.page.wait_for_load_state("networkidle", timeout=3000)
                if self.test_logger:
                    self.test_logger.debug("Page loaded - network idle")
                return True
            except Exception:
                pass
                
            return False
        except Exception as e:
            if self.test_logger:
                self.test_logger.debug(f"Page load check failed: {e}")
            return False
    
    def get_page_identifier(self) -> str:
        """Get page identifier."""
        return "Azure ML Studio"
    
    async def navigate_to_studio(self) -> None:
        """Navigate to Azure ML Studio."""
        await self.navigate_to(self.base_url)
    
    async def select_workspace(self, workspace_name: str) -> None:
        """Select a specific workspace."""
        if self.test_logger:
            self.test_logger.action(f"Selecting workspace: {workspace_name}")
        
        await self.click_element(self.WORKSPACE_SELECTOR)
        workspace_option = f"[data-testid='workspace-option'][data-workspace='{workspace_name}']"
        await self.click_element(workspace_option)
        await self.wait_for_navigation()
    
    # Navigation methods
    async def navigate_to_compute(self) -> None:
        """Navigate to the Compute section."""
        if self.test_logger:
            self.test_logger.action("Navigating to Compute section")
        
        await self.click_element(self.COMPUTE_NAV_ITEM)
        await self.wait_for_element(self.COMPUTE_INSTANCES_TAB)
    
    async def navigate_to_notebooks(self) -> None:
        """Navigate to the Notebooks section."""
        if self.test_logger:
            self.test_logger.action("Navigating to Notebooks section")
        
        await self.click_element(self.NOTEBOOKS_NAV_ITEM)
        await self.wait_for_element(self.NOTEBOOK_LIST)
    
    async def navigate_to_jobs(self) -> None:
        """Navigate to the Jobs section."""
        if self.test_logger:
            self.test_logger.action("Navigating to Jobs section")
        
        await self.click_element(self.JOBS_NAV_ITEM)
        await self.wait_for_element(self.JOB_LIST)
    
    async def navigate_to_datasets(self) -> None:
        """Navigate to the Datasets section."""
        if self.test_logger:
            self.test_logger.action("Navigating to Datasets section")
        
        await self.click_element(self.DATASETS_NAV_ITEM)
    
    # Compute Instance Management
    async def create_compute_instance(
        self,
        instance_name: str,
        vm_size: str = "Standard_DS3_v2"
    ) -> None:
        """Create a new compute instance."""
        if self.test_logger:
            self.test_logger.action(f"Creating compute instance: {instance_name}")
        
        # Navigate to compute instances tab
        await self.click_element(self.COMPUTE_INSTANCES_TAB)
        
        # Click create button
        await self.click_element(self.CREATE_COMPUTE_BUTTON)
        
        # Fill in instance details
        await self.fill_element(self.COMPUTE_INSTANCE_NAME_INPUT, instance_name)
        
        # Select VM size
        await self.click_element(self.COMPUTE_SIZE_DROPDOWN)
        vm_size_option = f"[data-testid='vm-size-option'][data-size='{vm_size}']"
        await self.click_element(vm_size_option)
        
        # Create the instance
        await self.click_element(self.CREATE_BUTTON)
        
        # Wait for creation to start
        await self.wait_for_element(self.COMPUTE_STATUS_CREATING)
        
        if self.test_logger:
            self.test_logger.info(f"Compute instance {instance_name} creation initiated")
    
    async def start_compute_instance(self, instance_name: str) -> None:
        """Start a compute instance."""
        if self.test_logger:
            self.test_logger.action(f"Starting compute instance: {instance_name}")
        
        instance_row = f"[data-testid='compute-instance-row'][data-name='{instance_name}']"
        await self.wait_for_element(instance_row)
        
        # Click start button for the specific instance
        start_button = f"{instance_row} {self.START_COMPUTE_BUTTON}"
        await self.click_element(start_button)
        
        if self.test_logger:
            self.test_logger.info(f"Start command sent for compute instance: {instance_name}")
    
    async def stop_compute_instance(self, instance_name: str) -> None:
        """Stop a compute instance."""
        if self.test_logger:
            self.test_logger.action(f"Stopping compute instance: {instance_name}")
        
        instance_row = f"[data-testid='compute-instance-row'][data-name='{instance_name}']"
        await self.wait_for_element(instance_row)
        
        # Click stop button for the specific instance
        stop_button = f"{instance_row} {self.STOP_COMPUTE_BUTTON}"
        await self.click_element(stop_button)
        
        if self.test_logger:
            self.test_logger.info(f"Stop command sent for compute instance: {instance_name}")
    
    async def delete_compute_instance(self, instance_name: str) -> None:
        """Delete a compute instance."""
        if self.test_logger:
            self.test_logger.action(f"Deleting compute instance: {instance_name}")
        
        instance_row = f"[data-testid='compute-instance-row'][data-name='{instance_name}']"
        await self.wait_for_element(instance_row)
        
        # Click delete button for the specific instance
        delete_button = f"{instance_row} {self.DELETE_COMPUTE_BUTTON}"
        await self.click_element(delete_button)
        
        # Confirm deletion
        confirm_button = "[data-testid='confirm-delete-button']"
        await self.click_element(confirm_button)
        
        if self.test_logger:
            self.test_logger.info(f"Delete command sent for compute instance: {instance_name}")
    
    async def get_compute_instance_status(self, instance_name: str) -> str:
        """Get the status of a compute instance."""
        instance_row = f"[data-testid='compute-instance-row'][data-name='{instance_name}']"
        await self.wait_for_element(instance_row)
        
        status_element = f"{instance_row} [data-testid='compute-status']"
        status = await self.get_element_text(status_element)
        
        if self.test_logger:
            self.test_logger.info(f"Compute instance {instance_name} status: {status}")
        
        return status
    
    async def wait_for_compute_instance_status(
        self,
        instance_name: str,
        expected_status: str,
        timeout: int = 600000  # 10 minutes
    ) -> None:
        """Wait for a compute instance to reach a specific status."""
        if self.test_logger:
            self.test_logger.action(
                f"Waiting for compute instance {instance_name} to reach status: {expected_status}"
            )
        
        async def check_status():
            current_status = await self.get_compute_instance_status(instance_name)
            return current_status.lower() == expected_status.lower()
        
        await self.wait_for_condition(check_status, timeout=timeout, poll_interval=10.0)
        
        if self.test_logger:
            self.test_logger.info(
                f"Compute instance {instance_name} reached status: {expected_status}"
            )
    
    async def assert_compute_instance_exists(self, instance_name: str) -> None:
        """Assert that a compute instance exists."""
        instance_row = f"[data-testid='compute-instance-row'][data-name='{instance_name}']"
        await self.assert_element_visible(instance_row)
        
        if self.test_logger:
            self.test_logger.assertion(f"Compute instance {instance_name} exists", True)
    
    # Notebook Management
    async def create_new_notebook(self, notebook_name: str) -> None:
        """Create a new notebook."""
        if self.test_logger:
            self.test_logger.action(f"Creating new notebook: {notebook_name}")
        
        await self.click_element(self.NEW_NOTEBOOK_BUTTON)
        
        # Fill notebook name
        name_input = "[data-testid='notebook-name-input']"
        await self.fill_element(name_input, notebook_name)
        
        # Create notebook
        create_button = "[data-testid='create-notebook-button']"
        await self.click_element(create_button)
        
        # Wait for notebook to load
        await self.wait_for_element(self.NOTEBOOK_CELL)
        
        if self.test_logger:
            self.test_logger.info(f"Notebook {notebook_name} created successfully")
    
    async def upload_notebook(self, file_path: str) -> None:
        """Upload a notebook file."""
        if self.test_logger:
            self.test_logger.action(f"Uploading notebook: {file_path}")
        
        await self.click_element(self.UPLOAD_NOTEBOOK_BUTTON)
        
        # Upload file
        file_input = "[data-testid='file-upload-input']"
        await self.upload_file(file_input, file_path)
        
        # Confirm upload
        upload_button = "[data-testid='confirm-upload-button']"
        await self.click_element(upload_button)
        
        if self.test_logger:
            self.test_logger.info(f"Notebook uploaded: {file_path}")
    
    async def open_notebook(self, notebook_name: str) -> None:
        """Open an existing notebook."""
        if self.test_logger:
            self.test_logger.action(f"Opening notebook: {notebook_name}")
        
        notebook_link = f"[data-testid='notebook-link'][data-name='{notebook_name}']"
        await self.click_element(notebook_link)
        
        # Wait for notebook to load
        await self.wait_for_element(self.NOTEBOOK_CELL)
        
        if self.test_logger:
            self.test_logger.info(f"Notebook {notebook_name} opened")
    
    async def run_notebook_cell(self, cell_index: int = 0) -> None:
        """Run a specific notebook cell."""
        if self.test_logger:
            self.test_logger.action(f"Running notebook cell {cell_index}")
        
        # Select the cell
        cell_selector = f"{self.NOTEBOOK_CELL}:nth-child({cell_index + 1})"
        await self.click_element(cell_selector)
        
        # Run the cell
        await self.click_element(self.RUN_CELL_BUTTON)
        
        if self.test_logger:
            self.test_logger.info(f"Notebook cell {cell_index} executed")
    
    async def run_all_notebook_cells(self) -> None:
        """Run all cells in the notebook."""
        if self.test_logger:
            self.test_logger.action("Running all notebook cells")
        
        await self.click_element(self.RUN_ALL_BUTTON)
        
        if self.test_logger:
            self.test_logger.info("All notebook cells executed")
    
    async def get_notebook_cell_output(self, cell_index: int = 0) -> str:
        """Get the output of a notebook cell."""
        cell_output = f"{self.NOTEBOOK_CELL}:nth-child({cell_index + 1}) [data-testid='cell-output']"
        await self.wait_for_element(cell_output)
        return await self.get_element_text(cell_output)
    
    # Job Management
    async def create_job(self, job_config: dict) -> None:
        """Create a new job."""
        if self.test_logger:
            self.test_logger.action("Creating new job")
        
        await self.click_element(self.CREATE_JOB_BUTTON)
        
        # Fill job configuration (simplified - actual implementation would be more complex)
        job_name_input = "[data-testid='job-name-input']"
        await self.fill_element(job_name_input, job_config.get("name", ""))
        
        # Submit job
        submit_button = "[data-testid='submit-job-button']"
        await self.click_element(submit_button)
        
        if self.test_logger:
            self.test_logger.info(f"Job {job_config.get('name')} created")
    
    async def get_job_status(self, job_name: str) -> str:
        """Get the status of a job."""
        job_row = f"[data-testid='job-row'][data-name='{job_name}']"
        await self.wait_for_element(job_row)
        
        status_element = f"{job_row} {self.JOB_STATUS}"
        status = await self.get_element_text(status_element)
        
        if self.test_logger:
            self.test_logger.info(f"Job {job_name} status: {status}")
        
        return status
    
    async def wait_for_job_completion(
        self,
        job_name: str,
        timeout: int = 1800000  # 30 minutes
    ) -> str:
        """Wait for a job to complete and return final status."""
        if self.test_logger:
            self.test_logger.action(f"Waiting for job {job_name} to complete")
        
        completed_statuses = ["completed", "failed", "canceled"]
        
        async def check_completion():
            status = await self.get_job_status(job_name)
            return status.lower() if status.lower() in completed_statuses else None
        
        final_status = await self.wait_for_condition(
            check_completion,
            timeout=timeout,
            poll_interval=30.0
        )
        
        if self.test_logger:
            self.test_logger.info(f"Job {job_name} completed with status: {final_status}")
        
        return final_status
    
    # Utility methods
    async def get_workspace_info(self) -> dict:
        """Get current workspace information."""
        workspace_info = {}
        
        # Get workspace name
        workspace_name_element = "[data-testid='workspace-name']"
        if await self.is_element_visible(workspace_name_element):
            workspace_info["name"] = await self.get_element_text(workspace_name_element)
        
        # Get subscription info
        subscription_element = "[data-testid='subscription-info']"
        if await self.is_element_visible(subscription_element):
            workspace_info["subscription"] = await self.get_element_text(subscription_element)
        
        return workspace_info
    
    async def assert_workspace_loaded(self, workspace_name: str) -> None:
        """Assert that the correct workspace is loaded."""
        workspace_info = await self.get_workspace_info()
        current_workspace = workspace_info.get("name", "")
        
        if self.test_logger:
            self.test_logger.assertion(
                f"Workspace {workspace_name} is loaded",
                current_workspace == workspace_name
            )
    
    # VS Code Desktop Integration
    async def launch_vscode_desktop(self, instance_name: str) -> None:
        """Launch VS Code Desktop for a compute instance."""
        if self.test_logger:
            self.test_logger.action(f"Launching VS Code Desktop for instance: {instance_name}")
        
        instance_row = f"[data-testid='compute-instance-row'][data-name='{instance_name}']"
        await self.wait_for_element(instance_row)
        
        # Try direct VS Code Desktop button first
        vscode_button = f"{instance_row} {self.VSCODE_DESKTOP_BUTTON}"
        if await self.is_element_visible(vscode_button, timeout=2000):
            await self.click_element(vscode_button)
            if self.test_logger:
                self.test_logger.info(f"VS Code Desktop launched directly for: {instance_name}")
            return
        
        # Try applications dropdown approach
        applications_button = f"{instance_row} {self.APPLICATIONS_BUTTON}"
        if await self.is_element_visible(applications_button, timeout=2000):
            await self.click_element(applications_button)
            
            # Wait for dropdown and select VS Code Desktop
            vscode_option = "[data-testid='application-option'][data-app='vscode-desktop']"
            await self.wait_for_element(vscode_option)
            await self.click_element(vscode_option)
            
            if self.test_logger:
                self.test_logger.info(f"VS Code Desktop launched via applications menu for: {instance_name}")
            return
        
        # Try alternative selectors
        alternative_selectors = [
            f"{instance_row} [title*='VS Code'][title*='Desktop']",
            f"{instance_row} [href*='vscode']",
            f"{instance_row} .application-link[title*='VS Code']"
        ]
        
        for selector in alternative_selectors:
            if await self.is_element_visible(selector, timeout=1000):
                await self.click_element(selector)
                if self.test_logger:
                    self.test_logger.info(f"VS Code Desktop launched via alternative selector for: {instance_name}")
                return
        
        raise Exception(f"Could not find VS Code Desktop launch option for instance: {instance_name}")
    
    async def wait_for_vscode_connection(self, timeout: int = 60000) -> bool:
        """Wait for VS Code Desktop connection to be established."""
        if self.test_logger:
            self.test_logger.action("Waiting for VS Code Desktop connection")
        
        # This is a placeholder implementation
        # In a real scenario, you might:
        # 1. Check for VS Code process
        # 2. Monitor connection status indicators
        # 3. Wait for specific UI elements
        
        import asyncio
        await asyncio.sleep(30)  # Give time for VS Code to start and connect
        
        if self.test_logger:
            self.test_logger.info("VS Code Desktop connection established (assumed)")
        
        return True
    
    # Enhanced Notebook Management
    async def add_notebook_cell_code(self, code: str, cell_index: int = 0) -> None:
        """Add code to a specific notebook cell."""
        if self.test_logger:
            self.test_logger.action(f"Adding code to notebook cell {cell_index}")
        
        # Select the cell
        cell_selector = f"{self.NOTEBOOK_CELL}:nth-child({cell_index + 1})"
        await self.click_element(cell_selector)
        
        # Clear existing content and add new code
        code_input = f"{cell_selector} [data-testid='cell-code-input']"
        await self.clear_element(code_input)
        await self.fill_element(code_input, code)
        
        if self.test_logger:
            self.test_logger.info(f"Code added to notebook cell {cell_index}")
    
    async def create_notebook_with_code(self, notebook_name: str, code_cells: list) -> None:
        """Create a new notebook with predefined code cells."""
        if self.test_logger:
            self.test_logger.action(f"Creating notebook with code: {notebook_name}")
        
        # Create the notebook
        await self.create_new_notebook(notebook_name)
        
        # Add code to cells
        for i, code in enumerate(code_cells):
            if i > 0:
                # Add new cell if needed
                add_cell_button = "[data-testid='add-cell-button']"
                await self.click_element(add_cell_button)
            
            await self.add_notebook_cell_code(code, i)
        
        if self.test_logger:
            self.test_logger.info(f"Notebook {notebook_name} created with {len(code_cells)} code cells")
    
    async def run_notebook_and_get_outputs(self) -> list:
        """Run all notebook cells and collect outputs."""
        if self.test_logger:
            self.test_logger.action("Running notebook and collecting outputs")
        
        await self.run_all_notebook_cells()
        
        # Wait for execution to complete
        await self.wait_for_notebook_execution_complete()
        
        # Collect outputs from all cells
        outputs = []
        cell_count = await self.get_notebook_cell_count()
        
        for i in range(cell_count):
            try:
                output = await self.get_notebook_cell_output(i)
                outputs.append(output)
            except Exception as e:
                if self.test_logger:
                    self.test_logger.warning(f"Could not get output from cell {i}: {e}")
                outputs.append(None)
        
        if self.test_logger:
            self.test_logger.info(f"Collected outputs from {len(outputs)} cells")
        
        return outputs
    
    async def wait_for_notebook_execution_complete(self, timeout: int = 300000) -> None:
        """Wait for notebook execution to complete."""
        if self.test_logger:
            self.test_logger.action("Waiting for notebook execution to complete")
        
        # Wait for all cells to finish executing
        async def check_execution_complete():
            executing_cells = await self.page.query_selector_all("[data-testid='cell-executing']")
            return len(executing_cells) == 0
        
        await self.wait_for_condition(check_execution_complete, timeout=timeout, poll_interval=2.0)
        
        if self.test_logger:
            self.test_logger.info("Notebook execution completed")
    
    async def get_notebook_cell_count(self) -> int:
        """Get the number of cells in the current notebook."""
        cells = await self.page.query_selector_all(self.NOTEBOOK_CELL)
        return len(cells)